{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing of bcftools vcf ('samtools')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import io\n",
    "import os\n",
    "import gzip\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plan:\n",
    "1. Variant calling\n",
    "2. Filtering\n",
    "3. Annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Variant calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "REF=/path/to/ref/S288C_reference_sequence_R64-3-1_20210421_chr.fasta\n",
    "BAMS=bams_list_LL13.txt\n",
    "bcftools mpileup -C50 -f ${REF} -min-MQ 4 -min-BQ 13 --skip-any-set 1796 -a FORMAT/AD,FORMAT/ADR,FORMAT/ADF,FORMAT/DP -Ou -b ${BAMS} | \\\n",
    "bcftools call --ploidy 1 -mv -f gq -Oz -o snp_bcftools_LL13.vcf.gz -\n",
    "tabix -p vcf snp_bcftools_LL13.vcf.gz\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "TMP=../DATA_temp\n",
    "SITES=../DATA/S288C_reference_sequence_R64-3-1_mod_noRepeats.bed\n",
    "\n",
    "for bg in LL13 NC02\n",
    "  do\n",
    "  echo $bg\n",
    "\n",
    "  # Sample filters\n",
    "  # 1. Masking genotypes with DP < 4 (mask) and GT 2 or 3; \n",
    "  # 2. Allelic depth (AD) of the second most common variant within genotype is less than 4\n",
    "  # 3. Allelic depth of the second most frequent variant to most frequent variant is less than 0.2\n",
    "\n",
    "  python filterAD2.py ../DATA/snp_bcftools_${bg}.vcf.gz\n",
    "  mv ../DATA/snp_bcftools_${bg}_filterAD.vcf $TMP/snp_bcftools_${bg}_filterAD.vcf\n",
    "  bgzip -f $TMP/snp_bcftools_${bg}_filterAD.vcf .\n",
    "  tabix -f -p vcf $TMP/snp_bcftools_${bg}_filterAD.vcf.gz\n",
    "\n",
    "  # SNP filters\n",
    "  # 1. Mean per sample coverge is higher than 10\n",
    "  # 2. Max total DP is less than 20000\n",
    "  # 3. SNP QUAL is higher than 20\n",
    "  # 4. Mapping quality MQ is higher than 40\n",
    "  # 5. Fixed alleles removed (AF>0.99 or AC==0)\n",
    "\n",
    "  bcftools filter -e \"AVG(FORMAT/DP)<10 || INFO/DP>20000 || QUAL<20 || MQ<40 || (AF > 0.99) || (AC == 0)\" -Ou $TMP/snp_bcftools_${bg}_filterAD.vcf.gz | bcftools view -m2 -Oz -o $TMP/snp_bcftools_${bg}_FLT4.vcf.gz -\n",
    "  bcftools stats $TMP/snp_bcftools_${bg}_FLT4.vcf.gz > snp_bcftools_${bg}_FLT4.stats\n",
    "  tabix -f -p vcf $TMP/snp_bcftools_${bg}_FLT4.vcf.gz\n",
    "  \n",
    "  done\n",
    "```\n",
    "Python script available here: [filterAD2_bcftools.py](filterAD2_bcftools.py)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Annotating with snpEff and *S. cerevisiae* genome vR64-3-1 and splitting into SNPs and INDELs\n",
    "```\n",
    "conda activate snpeff-5.0\n",
    "# SnpEff version SnpEff 5.0e (build 2021-03-09 06:01), by Pablo Cingolani\n",
    "snpEff_DIR=/home/anna/soft/snpEff\n",
    "TMP=../DATA_temp\n",
    "CONFIG=/home/software/snpEff/snpEff.config\n",
    "\n",
    "for bg in LL13 NC02\n",
    "  do\n",
    "\n",
    "  # SnpEff\n",
    "  snpEff -c $CONFIG R64.3.1 $TMP/snp_bcftools_${bg}_FLT4.vcf.gz > $TMP/snp_bcftools_${bg}_snpEff.vcf\n",
    "  bgzip -f $TMP/snp_bcftools_${bg}_snpEff.vcf\n",
    "  tabix -f -p vcf $TMP/snp_bcftools_${bg}_snpEff.vcf.gz\n",
    "\n",
    "  # Saving table\n",
    "  bcftools query -i '%TYPE=\"snp\"' -f '%CHROM\\t%POS\\t%REF\\t%ALT\\t%AC\\t\\t%ANN\\n' -o snp_bcftools_${bg}_SNP_snpEff.tab $TMP/snp_bcftools_${bg}_snpEff.vcf.gz\n",
    "  bcftools query -i '%TYPE=\"indel\"' -f '%CHROM\\t%POS\\t%REF\\t%ALT\\t%AC\\t\\t%ANN\\n' -o snp_bcftools_${bg}_INDEL_snpEff.tab $TMP/snp_bcftools_${bg}_snpEff.vcf.gz\n",
    "  cp $TMP/snp_bcftools_${bg}_snpEff.vcf.gz* ../DATA/\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Getting a summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "important_categories = ['synonymous_variant','stop_gained','missense_variant']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Function for reading vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readVcfGzip(filepath):\n",
    "    fh = gzip.open(filepath,'rt')\n",
    "    lines = [l for l in fh if not l.startswith('##')]\n",
    "    df_vcf = pd.read_csv(io.StringIO(''.join(lines)),sep='\\t')\n",
    "    return(df_vcf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Function for formatting vcf into stacked table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stackMutations(vcf_file):\n",
    "    \n",
    "    # Vcf to dataframe\n",
    "    df_vcf = readVcfGzip(vcf_file)\n",
    "    \n",
    "    # Formatting snpeff annotations & mutation\n",
    "    infos = df_vcf['INFO'].values.tolist()\n",
    "    I, G = [],[]\n",
    "    #T = []\n",
    "    for info in infos:\n",
    "        muts = []\n",
    "        ninfo = [i.split('=')[1] for i in info.split(';') if i.startswith('ANN=')][0]\n",
    "        sinfo = [i for i in ninfo.split(',')]\n",
    "        #T.append(ninfo)\n",
    "        mut = \":\".join([sinfo[0].split('|')[0],sinfo[0].split('|')[1]]), sinfo[0].split('|')[3].replace(\"_CDS\",\"\")\n",
    "        muts.append(mut)\n",
    "        for gene in sinfo[1:]:\n",
    "            ginfo = gene.split('|')\n",
    "            mut = \":\".join([ginfo[0],ginfo[1]]), ginfo[3].replace(\"_CDS\",\"\")\n",
    "            if (ginfo[1] in important_categories) & (mut not in muts):\n",
    "                muts.append(mut)\n",
    "        all_muts = '/'.join([m[0] for m in muts])\n",
    "        all_genes = '/'.join([m[1] for m in muts])\n",
    "        I.append(all_muts)\n",
    "        G.append(all_genes)\n",
    "    df_vcf['MUT'] = I\n",
    "    df_vcf['GENE'] = G\n",
    "    df_vcf['mutID'] = df_vcf.apply(lambda x: x[\"#CHROM\"]+\":\"+str(x[\"POS\"])+\":\"+str(x[\"REF\"])+\":\"+x[\"MUT\"], axis=1)\n",
    "    df_vcf.head()\n",
    "    \n",
    "    # Formatting samples\n",
    "    df_vcf_ = df_vcf.drop(columns=[\"ID\",\"FILTER\",\"INFO\",\"FORMAT\",\"MUT\"])\n",
    "    df_stack = df_vcf_.set_index([\"#CHROM\",\"POS\",\"REF\",\"ALT\",\"QUAL\",\"GENE\",\"mutID\"]).stack().reset_index().rename(columns={\"level_7\":\"sampleID\",0:\"GT\"})\n",
    "    del df_vcf_\n",
    "    df_stack['sample'] = df_stack['sampleID'].apply(lambda x: x.split('/')[-1].split('.')[0])\n",
    "    df_stack['snp'] = df_stack['GT'].apply(lambda x: 1 if x.split(':')[0] == '1' else 0)\n",
    "    df_stack = df_stack.drop(columns=[\"REF\",\"ALT\",\"QUAL\",\"sampleID\",\"GT\"])\n",
    "    df_stack = df_stack[df_stack['snp']==1].reset_index(drop=True)\n",
    "    del df_vcf\n",
    "    return(df_stack)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stackMutationsWithAnn(vcf_file):\n",
    "    \n",
    "    # Vcf to dataframe\n",
    "    df_vcf = readVcfGzip(vcf_file)\n",
    "    \n",
    "    # Formatting snpeff annotations & mutation\n",
    "    infos = df_vcf['INFO'].values.tolist()\n",
    "    I, G = [],[]\n",
    "    #T = []\n",
    "    for info in infos:\n",
    "        muts = []\n",
    "        ninfo = [i.split('=')[1] for i in info.split(';') if i.startswith('ANN=')][0]\n",
    "        sinfo = [i for i in ninfo.split(',')]\n",
    "        #T.append(ninfo)\n",
    "        mut = \":\".join([sinfo[0].split('|')[0],sinfo[0].split('|')[1]]), sinfo[0].split('|')[3].replace(\"_CDS\",\"\")\n",
    "        muts.append(mut)\n",
    "        for gene in sinfo[1:]:\n",
    "            ginfo = gene.split('|')\n",
    "            mut = \":\".join([ginfo[0],ginfo[1]]), ginfo[3].replace(\"_CDS\",\"\")\n",
    "            if (ginfo[1] in important_categories) & (mut not in muts):\n",
    "                muts.append(mut)\n",
    "        all_muts = '/'.join([m[0] for m in muts])\n",
    "        all_genes = '/'.join([m[1] for m in muts])\n",
    "        I.append(all_muts)\n",
    "        G.append(all_genes)\n",
    "        \n",
    "    A = []\n",
    "    ANN = []\n",
    "    for info in infos:\n",
    "        muts = []\n",
    "        ninfo = [i.split('=')[1] for i in info.split(';') if i.startswith('ANN=')][0]\n",
    "        ANN.append(ninfo)\n",
    "        sinfo = [i for i in ninfo.split(',')]\n",
    "        #mut = \":\".join([sinfo[0].split('|')[0],sinfo[0].split('|')[1]]), sinfo[0].split('|')[3].replace(\"_CDS\",\"\")\n",
    "        #muts.append(mut)\n",
    "        for gene in sinfo:\n",
    "            ginfo = gene.split('|')\n",
    "            mut = \":\".join([ginfo[0],ginfo[1],ginfo[3].replace(\"_CDS\",\"\")])\n",
    "            if mut not in muts:\n",
    "                muts.append(mut)\n",
    "        all_muts = '/'.join(muts)\n",
    "        A.append(all_muts)\n",
    "    \n",
    "    df_vcf['MUT'] = I\n",
    "    df_vcf['GENE'] = G\n",
    "    df_vcf['MUT_ALL'] = A\n",
    "    df_vcf['ANN'] = ANN\n",
    "    df_vcf['mutID'] = df_vcf.apply(lambda x: x[\"#CHROM\"]+\":\"+str(x[\"POS\"])+\":\"+str(x[\"REF\"])+\":\"+x[\"MUT\"], axis=1)\n",
    "    df_vcf.head()\n",
    "    \n",
    "    # Formatting samples\n",
    "    df_vcf_ = df_vcf.drop(columns=[\"ID\",\"FILTER\",\"INFO\",\"FORMAT\",\"MUT\"])\n",
    "    df_stack = df_vcf_.set_index([\"#CHROM\",\"POS\",\"REF\",\"ALT\",\"QUAL\",\"GENE\",\"mutID\",\"MUT_ALL\",\"ANN\"]).stack().reset_index().rename(columns={\"level_9\":\"sampleID\",0:\"GT\"})\n",
    "    del df_vcf_\n",
    "    df_stack['sample'] = df_stack['sampleID'].apply(lambda x: x.split('/')[-1].split('.')[0])\n",
    "    df_stack['snp'] = df_stack['GT'].apply(lambda x: 1 if x.split(':')[0] == '1' else 0)\n",
    "    df_stack = df_stack.drop(columns=[\"QUAL\",\"sampleID\"])\n",
    "    df_stack = df_stack[df_stack['snp']==1].reset_index(drop=True)\n",
    "    del df_vcf\n",
    "    return(df_stack)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combining two backgrounds together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4348, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:42,0:4:0,1:0,3:0,4:41</td>\n",
       "      <td>F4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:86,0:8:0,3:0,5:0,8:85</td>\n",
       "      <td>F5</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:55,0:8:1,6:0,1:1,7:54</td>\n",
       "      <td>J2</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:45,1:7:0,5:1,1:1,6:43</td>\n",
       "      <td>L2</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:63,0:7:0,2:1,3:1,5:62</td>\n",
       "      <td>M4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM    POS REF ALT     GENE                              mutID  \\\n",
       "0   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "1   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "2   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "3   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "4   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "1  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "2  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "3  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "4  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "\n",
       "                                                 ANN                       GT  \\\n",
       "0  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:42,0:4:0,1:0,3:0,4:41   \n",
       "1  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:86,0:8:0,3:0,5:0,8:85   \n",
       "2  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:55,0:8:1,6:0,1:1,7:54   \n",
       "3  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:45,1:7:0,5:1,1:1,6:43   \n",
       "4  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:63,0:7:0,2:1,3:1,5:62   \n",
       "\n",
       "  sample  snp background     _merge  \n",
       "0     F4    1   LL13-040  left_only  \n",
       "1     F5    1   LL13-040  left_only  \n",
       "2     J2    1   LL13-040  left_only  \n",
       "3     L2    1   LL13-040  left_only  \n",
       "4     M4    1   LL13-040  left_only  "
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf1 = \"../DATA/snp_bcftools_LL13_snpEff_SNP.vcf.gz\"\n",
    "vcf2 = \"../DATA/snp_bcftools_NC02_snpEff_SNP.vcf.gz\"\n",
    "# LL13\n",
    "df_vcf1 = stackMutationsWithAnn(vcf1)\n",
    "df_vcf1['background'] = \"LL13-040\"\n",
    "parent_variants_LL13 = df_vcf1.loc[df_vcf1['sample'].isin(['H13']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf1_merged = pd.merge(df_vcf1, parent_variants_LL13, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf1_noWT = df_vcf1_merged[df_vcf1_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "# NC02\n",
    "df_vcf2 = stackMutationsWithAnn(vcf2)\n",
    "df_vcf2['background'] = \"NC-02\"\n",
    "parent_variants_NC02 = df_vcf2.loc[df_vcf2['sample'].isin(['E23']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf2_merged = pd.merge(df_vcf2, parent_variants_NC02, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf2_noWT = df_vcf2_merged[df_vcf2_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "\n",
    "# Merged\n",
    "df = pd.concat([df_vcf1_noWT, df_vcf2_noWT],ignore_index=True)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['NGENE'] = df['GENE'].apply(lambda x: list(set(x.split('/')))[0] if len(set(x.split('/'))) == 1 else x)\n",
    "#df[df['GENE'] == \"YHR128W/YHR128W\"].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Getting gene descriptions from SGD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading gene names extracted from SGD for corresponding gene IDs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Species</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S000000001</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>TFC3</td>\n",
       "      <td>Transcription Factor class C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S000000002</td>\n",
       "      <td>YAL002W</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>VPS8</td>\n",
       "      <td>Vacuolar Protein Sorting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S000000003</td>\n",
       "      <td>YAL003W</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>EFB1</td>\n",
       "      <td>Elongation Factor Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S000000004</td>\n",
       "      <td>YAL005C</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>SSA1</td>\n",
       "      <td>Stress-Seventy subfamily A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S000000005</td>\n",
       "      <td>YAL007C</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>ERP2</td>\n",
       "      <td>Emp24p/Erv25p Related Protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID     Gene        Species Symbol                           Name\n",
       "0  S000000001  YAL001C  S. cerevisiae   TFC3   Transcription Factor class C\n",
       "1  S000000002  YAL002W  S. cerevisiae   VPS8       Vacuolar Protein Sorting\n",
       "2  S000000003  YAL003W  S. cerevisiae   EFB1         Elongation Factor Beta\n",
       "3  S000000004  YAL005C  S. cerevisiae   SSA1     Stress-Seventy subfamily A\n",
       "4  S000000005  YAL007C  S. cerevisiae   ERP2  Emp24p/Erv25p Related Protein"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dg = pd.read_csv('../../DATA/S288C_reference_genome_R64-3-1_20210421/gene_list_SGD.tsv',sep=\"\\t\",names=['ID','Gene','Species','Symbol','Name'])\n",
    "dg.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Adding missing gene IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(4348, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:42,0:4:0,1:0,3:0,4:41</td>\n",
       "      <td>F4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>FUN12</td>\n",
       "      <td>Function Unknown Now</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:86,0:8:0,3:0,5:0,8:85</td>\n",
       "      <td>F5</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>FUN12</td>\n",
       "      <td>Function Unknown Now</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:55,0:8:1,6:0,1:1,7:54</td>\n",
       "      <td>J2</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>FUN12</td>\n",
       "      <td>Function Unknown Now</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:45,1:7:0,5:1,1:1,6:43</td>\n",
       "      <td>L2</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>FUN12</td>\n",
       "      <td>Function Unknown Now</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>77527</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>chrI:77527:A:G:synonymous_variant</td>\n",
       "      <td>G:synonymous_variant:YAL035W/G:upstream_gene_v...</td>\n",
       "      <td>G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...</td>\n",
       "      <td>1:63,0:7:0,2:1,3:1,5:62</td>\n",
       "      <td>M4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL035W</td>\n",
       "      <td>FUN12</td>\n",
       "      <td>Function Unknown Now</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM    POS REF ALT     GENE                              mutID  \\\n",
       "0   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "1   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "2   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "3   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "4   chrI  77527   A   G  YAL035W  chrI:77527:A:G:synonymous_variant   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "1  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "2  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "3  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "4  G:synonymous_variant:YAL035W/G:upstream_gene_v...   \n",
       "\n",
       "                                                 ANN                       GT  \\\n",
       "0  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:42,0:4:0,1:0,3:0,4:41   \n",
       "1  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:86,0:8:0,3:0,5:0,8:85   \n",
       "2  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:55,0:8:1,6:0,1:1,7:54   \n",
       "3  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:45,1:7:0,5:1,1:1,6:43   \n",
       "4  G|synonymous_variant|LOW|YAL035W_CDS|GENE_YAL0...  1:63,0:7:0,2:1,3:1,5:62   \n",
       "\n",
       "  sample  snp background     _merge    NGENE GENE_SYMBOL             GENE_NAME  \n",
       "0     F4    1   LL13-040  left_only  YAL035W       FUN12  Function Unknown Now  \n",
       "1     F5    1   LL13-040  left_only  YAL035W       FUN12  Function Unknown Now  \n",
       "2     J2    1   LL13-040  left_only  YAL035W       FUN12  Function Unknown Now  \n",
       "3     L2    1   LL13-040  left_only  YAL035W       FUN12  Function Unknown Now  \n",
       "4     M4    1   LL13-040  left_only  YAL035W       FUN12  Function Unknown Now  "
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dictG = {ele:np.array(ele.split('/')) for ele in df['NGENE'].drop_duplicates().values}\n",
    "all_gene_ids = np.concatenate(list(dictG.values()))\n",
    "missing_IDs = [ele for ele in all_gene_ids if ele not in dg['Gene'].values]\n",
    "missing_dg = pd.DataFrame({\"ID\":\"SXX\",\"Gene\":missing_IDs,\"Species\":\"S. cerevisiae\",\"Symbol\":\"UNK\",\"Name\":\"UNK\"})\n",
    "dg_combined = pd.concat([dg, missing_dg], ignore_index = True)\n",
    "\n",
    "gene_symbol = {a:b for a,b in dg_combined.get(['Gene','Symbol']).fillna('UNK').values.tolist()}\n",
    "gene_name = {a:b for a,b in dg_combined.get(['Gene','Name']).fillna('UNK').values.tolist()}\n",
    "\n",
    "df['GENE_SYMBOL'] = df['NGENE'].apply(lambda x: '/'.join([gene_symbol[i] for i in x.split('/')]))\n",
    "df['GENE_NAME'] = df['NGENE'].apply(lambda x: '/'.join([gene_name[i] for i in x.split('/')]))\n",
    "print(df.shape)\n",
    "df.to_csv('samtools_variants_SNP.tab',sep='\\t',header=True,index=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# INDELs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combining two backgrounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8461, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>6736</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...</td>\n",
       "      <td>1:106,0,101:13:0,8,1:1,3,0:1,11,1:127</td>\n",
       "      <td>B13</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>6736</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...</td>\n",
       "      <td>1:87,0,101:8:1,7,0:0,0,0:1,7,0:127</td>\n",
       "      <td>F12</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>13051</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGACTAGAGAAGAAG...</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>chrI:13051:GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGAC...</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT:upstream_gene...</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT|upstream_gene...</td>\n",
       "      <td>1:119,11:7:1,0:0,6:1,6:109</td>\n",
       "      <td>B11</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>31500</td>\n",
       "      <td>CACACATATATA</td>\n",
       "      <td>CATATA</td>\n",
       "      <td>YAL063C</td>\n",
       "      <td>chrI:31500:CACACATATATA:CATATA:upstream_gene_v...</td>\n",
       "      <td>CATATA:upstream_gene_variant:YAL063C/CATATA:up...</td>\n",
       "      <td>CATATA|upstream_gene_variant|MODIFIER|YAL063C_...</td>\n",
       "      <td>1:150,71:16:2,7:0,4:2,11:76</td>\n",
       "      <td>B6</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL063C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>31500</td>\n",
       "      <td>CACACATATATA</td>\n",
       "      <td>CATATA</td>\n",
       "      <td>YAL063C</td>\n",
       "      <td>chrI:31500:CACACATATATA:CATATA:upstream_gene_v...</td>\n",
       "      <td>CATATA:upstream_gene_variant:YAL063C/CATATA:up...</td>\n",
       "      <td>CATATA|upstream_gene_variant|MODIFIER|YAL063C_...</td>\n",
       "      <td>1:104,0:13:1,9:1,1:2,10:101</td>\n",
       "      <td>C12</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL063C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM    POS                                                REF  \\\n",
       "0   chrI   6736                               CAAAAAAAAAAAAAAAAAAA   \n",
       "1   chrI   6736                               CAAAAAAAAAAAAAAAAAAA   \n",
       "2   chrI  13051  GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGACTAGAGAAGAAG...   \n",
       "3   chrI  31500                                       CACACATATATA   \n",
       "4   chrI  31500                                       CACACATATATA   \n",
       "\n",
       "                                            ALT     GENE  \\\n",
       "0  CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA  YAL068C   \n",
       "1  CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA  YAL068C   \n",
       "2              GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT  YAL067C   \n",
       "3                                        CATATA  YAL063C   \n",
       "4                                        CATATA  YAL063C   \n",
       "\n",
       "                                               mutID  \\\n",
       "0  chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...   \n",
       "1  chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...   \n",
       "2  chrI:13051:GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGAC...   \n",
       "3  chrI:31500:CACACATATATA:CATATA:upstream_gene_v...   \n",
       "4  chrI:31500:CACACATATATA:CATATA:upstream_gene_v...   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...   \n",
       "1  CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...   \n",
       "2  GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT:upstream_gene...   \n",
       "3  CATATA:upstream_gene_variant:YAL063C/CATATA:up...   \n",
       "4  CATATA:upstream_gene_variant:YAL063C/CATATA:up...   \n",
       "\n",
       "                                                 ANN  \\\n",
       "0  CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...   \n",
       "1  CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...   \n",
       "2  GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT|upstream_gene...   \n",
       "3  CATATA|upstream_gene_variant|MODIFIER|YAL063C_...   \n",
       "4  CATATA|upstream_gene_variant|MODIFIER|YAL063C_...   \n",
       "\n",
       "                                      GT sample  snp background     _merge  \\\n",
       "0  1:106,0,101:13:0,8,1:1,3,0:1,11,1:127    B13    1   LL13-040  left_only   \n",
       "1     1:87,0,101:8:1,7,0:0,0,0:1,7,0:127    F12    1   LL13-040  left_only   \n",
       "2             1:119,11:7:1,0:0,6:1,6:109    B11    1   LL13-040  left_only   \n",
       "3            1:150,71:16:2,7:0,4:2,11:76     B6    1   LL13-040  left_only   \n",
       "4            1:104,0:13:1,9:1,1:2,10:101    C12    1   LL13-040  left_only   \n",
       "\n",
       "     NGENE  \n",
       "0  YAL068C  \n",
       "1  YAL068C  \n",
       "2  YAL067C  \n",
       "3  YAL063C  \n",
       "4  YAL063C  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf1 = \"../DATA/snp_bcftools_LL13_snpEff_INDEL.vcf.gz\"\n",
    "vcf2 = \"../DATA/snp_bcftools_NC02_snpEff_INDEL.vcf.gz\"\n",
    "# LL13\n",
    "df_vcf1 = stackMutationsWithAnn(vcf1)\n",
    "df_vcf1['background'] = \"LL13-040\"\n",
    "parent_variants_LL13 = df_vcf1.loc[df_vcf1['sample'].isin(['H13']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf1_merged = pd.merge(df_vcf1, parent_variants_LL13, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf1_noWT = df_vcf1_merged[df_vcf1_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "# NC02\n",
    "df_vcf2 = stackMutationsWithAnn(vcf2)\n",
    "df_vcf2['background'] = \"NC-02\"\n",
    "parent_variants_NC02 = df_vcf2.loc[df_vcf2['sample'].isin(['E23']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf2_merged = pd.merge(df_vcf2, parent_variants_NC02, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf2_noWT = df_vcf2_merged[df_vcf2_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "\n",
    "# Merged\n",
    "df = pd.concat([df_vcf1_noWT, df_vcf2_noWT],ignore_index=True)\n",
    "print(df.shape)\n",
    "\n",
    "# Gene name\n",
    "df['NGENE'] = df['GENE'].apply(lambda x: list(set(x.split('/')))[0] if len(set(x.split('/'))) == 1 else x)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [#CHROM, POS, REF, ALT, GENE, mutID, MUT_ALL, ANN, GT, sample, snp, background, _merge, NGENE]\n",
       "Index: []"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[\"#CHROM\",\"POS\"]].drop_duplicates()\n",
    "df[df['NGENE'] == \"YHR128W\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SGD gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(8461, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>6736</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...</td>\n",
       "      <td>1:106,0,101:13:0,8,1:1,3,0:1,11,1:127</td>\n",
       "      <td>B13</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>seriPAUperin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>6736</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...</td>\n",
       "      <td>CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...</td>\n",
       "      <td>1:87,0,101:8:1,7,0:0,0,0:1,7,0:127</td>\n",
       "      <td>F12</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>seriPAUperin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>13051</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGACTAGAGAAGAAG...</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>chrI:13051:GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGAC...</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT:upstream_gene...</td>\n",
       "      <td>GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT|upstream_gene...</td>\n",
       "      <td>1:119,11:7:1,0:0,6:1,6:109</td>\n",
       "      <td>B11</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>SEO1</td>\n",
       "      <td>Suppressor of sulfoxyde EthiOnine resistance</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>31500</td>\n",
       "      <td>CACACATATATA</td>\n",
       "      <td>CATATA</td>\n",
       "      <td>YAL063C</td>\n",
       "      <td>chrI:31500:CACACATATATA:CATATA:upstream_gene_v...</td>\n",
       "      <td>CATATA:upstream_gene_variant:YAL063C/CATATA:up...</td>\n",
       "      <td>CATATA|upstream_gene_variant|MODIFIER|YAL063C_...</td>\n",
       "      <td>1:150,71:16:2,7:0,4:2,11:76</td>\n",
       "      <td>B6</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL063C</td>\n",
       "      <td>FLO9</td>\n",
       "      <td>FLOcculation</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>31500</td>\n",
       "      <td>CACACATATATA</td>\n",
       "      <td>CATATA</td>\n",
       "      <td>YAL063C</td>\n",
       "      <td>chrI:31500:CACACATATATA:CATATA:upstream_gene_v...</td>\n",
       "      <td>CATATA:upstream_gene_variant:YAL063C/CATATA:up...</td>\n",
       "      <td>CATATA|upstream_gene_variant|MODIFIER|YAL063C_...</td>\n",
       "      <td>1:104,0:13:1,9:1,1:2,10:101</td>\n",
       "      <td>C12</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL063C</td>\n",
       "      <td>FLO9</td>\n",
       "      <td>FLOcculation</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM    POS                                                REF  \\\n",
       "0   chrI   6736                               CAAAAAAAAAAAAAAAAAAA   \n",
       "1   chrI   6736                               CAAAAAAAAAAAAAAAAAAA   \n",
       "2   chrI  13051  GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGACTAGAGAAGAAG...   \n",
       "3   chrI  31500                                       CACACATATATA   \n",
       "4   chrI  31500                                       CACACATATATA   \n",
       "\n",
       "                                            ALT     GENE  \\\n",
       "0  CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA  YAL068C   \n",
       "1  CAAAAAAAAAAAAAAAAAAAA,CAAAAAAAAAAAAAAAAAAAAA  YAL068C   \n",
       "2              GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT  YAL067C   \n",
       "3                                        CATATA  YAL063C   \n",
       "4                                        CATATA  YAL063C   \n",
       "\n",
       "                                               mutID  \\\n",
       "0  chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...   \n",
       "1  chrI:6736:CAAAAAAAAAAAAAAAAAAA:CAAAAAAAAAAAAAA...   \n",
       "2  chrI:13051:GTGACTAGAGAAGAAGTGACTGAGGAAGAAATGAC...   \n",
       "3  chrI:31500:CACACATATATA:CATATA:upstream_gene_v...   \n",
       "4  chrI:31500:CACACATATATA:CATATA:upstream_gene_v...   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...   \n",
       "1  CAAAAAAAAAAAAAAAAAAAA:upstream_gene_variant:YA...   \n",
       "2  GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT:upstream_gene...   \n",
       "3  CATATA:upstream_gene_variant:YAL063C/CATATA:up...   \n",
       "4  CATATA:upstream_gene_variant:YAL063C/CATATA:up...   \n",
       "\n",
       "                                                 ANN  \\\n",
       "0  CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...   \n",
       "1  CAAAAAAAAAAAAAAAAAAAA|upstream_gene_variant|MO...   \n",
       "2  GTGACTAGAGAAGAAGTGTCTGAGGAAGAAAT|upstream_gene...   \n",
       "3  CATATA|upstream_gene_variant|MODIFIER|YAL063C_...   \n",
       "4  CATATA|upstream_gene_variant|MODIFIER|YAL063C_...   \n",
       "\n",
       "                                      GT sample  snp background     _merge  \\\n",
       "0  1:106,0,101:13:0,8,1:1,3,0:1,11,1:127    B13    1   LL13-040  left_only   \n",
       "1     1:87,0,101:8:1,7,0:0,0,0:1,7,0:127    F12    1   LL13-040  left_only   \n",
       "2             1:119,11:7:1,0:0,6:1,6:109    B11    1   LL13-040  left_only   \n",
       "3            1:150,71:16:2,7:0,4:2,11:76     B6    1   LL13-040  left_only   \n",
       "4            1:104,0:13:1,9:1,1:2,10:101    C12    1   LL13-040  left_only   \n",
       "\n",
       "     NGENE GENE_SYMBOL                                     GENE_NAME  \n",
       "0  YAL068C        PAU8                                  seriPAUperin  \n",
       "1  YAL068C        PAU8                                  seriPAUperin  \n",
       "2  YAL067C        SEO1  Suppressor of sulfoxyde EthiOnine resistance  \n",
       "3  YAL063C        FLO9                                  FLOcculation  \n",
       "4  YAL063C        FLO9                                  FLOcculation  "
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dg = pd.read_csv('../../DATA/S288C_reference_genome_R64-3-1_20210421/gene_list_SGD.tsv',sep=\"\\t\",names=['ID','Gene','Species','Symbol','Name'])\n",
    "dictG = {ele:np.array(ele.split('/')) for ele in df['NGENE'].drop_duplicates().values}\n",
    "all_gene_ids = np.concatenate(list(dictG.values()))\n",
    "missing_IDs = [ele for ele in all_gene_ids if ele not in dg['Gene'].values]\n",
    "missing_dg = pd.DataFrame({\"ID\":\"SXX\",\"Gene\":missing_IDs,\"Species\":\"S. cerevisiae\",\"Symbol\":\"UNK\",\"Name\":\"UNK\"})\n",
    "dg_combined = pd.concat([dg, missing_dg], ignore_index = True)\n",
    "\n",
    "gene_symbol = {a:b for a,b in dg_combined.get(['Gene','Symbol']).fillna('UNK').values.tolist()}\n",
    "gene_name = {a:b for a,b in dg_combined.get(['Gene','Name']).fillna('UNK').values.tolist()}\n",
    "\n",
    "df['GENE_SYMBOL'] = df['NGENE'].apply(lambda x: '/'.join([gene_symbol[i] for i in x.split('/')]))\n",
    "df['GENE_NAME'] = df['NGENE'].apply(lambda x: '/'.join([gene_name[i] for i in x.split('/')]))\n",
    "print(df.shape)\n",
    "df.to_csv('samtools_variants_INDEL.tab',sep='\\t',header=True,index=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
