{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Processing of gatk vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import io\n",
    "import os\n",
    "import gzip\n",
    "import sys"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Plan:\n",
    "1. Variant calling\n",
    "2. Filtering\n",
    "3. Annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1. Variant calling"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "Readjusting RG in bams\n",
    "java -jar picard.jar AddOrReplaceReadGroups I=${DIR}/bam_files/B10.sorted.calmd.bam O=${DIR}/bam_files_RG/B10_rg.bam RGID=HNVJGDSX2.4 RGLB=B10 RGPL=ILLUMINA RGPU=HNVJGDSX2.4.B10 RGSM=B10\n",
    "Running HaplotypeCaller\n",
    "gatk HaplotypeCaller -R ${DIR}/reference_files/S288C_reference_sequence_R64-3-1_20210421_chr.fasta -I ${DIR}/bam_files_RG/B10_rg.bam -O HaplotypeCaller/B10.g.vcf.gz -ploidy 1 -ERC GVCF --min-base-quality-score 20 --annotation StrandBiasBySample\n",
    "Running CombineGVCFs\n",
    "\n",
    "gatk CombineGVCFs -R $REF hapcalls_NC02.txt \\\n",
    " -O CombineGenotypes/set_NC02.g.vcf \\\n",
    " -G StandardAnnotation\n",
    "Running GenotypeGVCFs\n",
    "\n",
    "gatk GenotypeGVCFs -R $REF \\\n",
    "-V CombineGenotypes/set_LL13.g.vcf \\\n",
    "-O CombineGenotypes/set_LL13.vcf.gz \\\n",
    "-G StandardAnnotation \\\n",
    "--annotation Coverage \\\n",
    "--annotation QualByDepth \\\n",
    "--annotation MappingQuality \\\n",
    "--annotation MappingQualityRankSumTest \\\n",
    "--annotation ExcessHet\n",
    "Split to biallelic SNPs and INDELs\n",
    "\n",
    "gatk SelectVariants -R $REF \\\n",
    "-V CombineGenotypes/set_LL13.vcf.gz \\\n",
    "-O CombineGenotypes/set_LL13_SNPs.vcf.gz \\\n",
    "--restrict-alleles-to BIALLELIC \\\n",
    "--exclude-filtered true \\\n",
    "--exclude-non-variants true \\\n",
    "--remove-unused-alternates true \\\n",
    "--select-type SNP\n",
    "VariantFiltration for SNPs\n",
    "\n",
    "gatk VariantFiltration -R $REF \\\n",
    "-V CombineGenotypes/set_${set}_SNPs.vcf.gz \\\n",
    "-O CombineGenotypes/set_${set}_SNPs_VF.vcf.gz \\\n",
    "-filter \"vc.hasAttribute('QD') && QD < 2.0\" \\\n",
    "--filter-name \"QD\" \\\n",
    "-filter \"vc.hasAttribute('QUAL') && QUAL < 30.0\" \\\n",
    "--filter-name \"QUAL\" \\\n",
    "-filter \"vc.hasAttribute('MQ') && MQ < 40.0\" \\\n",
    "--filter-name \"MapQ\" \\\n",
    "-filter \"vc.hasAttribute('FS') && FS > 60.0\" \\\n",
    "--filter-name \"FS\" \\\n",
    "-filter \"vc.hasAttribute('FS') && SOR > 4.0\" \\\n",
    "--filter-name \"SOR\" \\\n",
    "-filter \"vc.hasAttribute('MQRankSum') && MQRankSum < -12.5\" \\\n",
    "--filter-name \"MQRS\" \\\n",
    "-filter \"vc.hasAttribute('ReadPosRankSum') && ReadPosRankSum < -8.0\" \\\n",
    "--filter-name \"RPRS\" \\\n",
    "--genotype-filter-expression \"DP < 4\" \\\n",
    "--genotype-filter-name \"DP\" \\\n",
    "--set-filtered-genotype-to-no-call true \\\n",
    "--QUIET true\n",
    "VariantFiltration for INDELs\n",
    "\n",
    "gatk VariantFiltration -R $REF \\\n",
    "-V CombineGenotypes/set_${set}_INDELs.vcf.gz \\\n",
    "-O CombineGenotypes/set_${set}_INDELs_VF.vcf.gz \\\n",
    "-filter \"vc.hasAttribute('QD') && QD < 2.0\" \\\n",
    "--filter-name \"QD\" \\\n",
    "-filter \"vc.hasAttribute('QUAL') && QUAL < 30.0\" \\\n",
    "--filter-name \"QUAL\" \\\n",
    "-filter \"vc.hasAttribute('FS') && FS > 200.0\" \\\n",
    "--filter-name \"FS\" \\\n",
    "-filter \"vc.hasAttribute('ReadPosRankSum') && ReadPosRankSum < -20.0\" \\\n",
    "--filter-name \"RPRS\" \\\n",
    "--genotype-filter-expression \"DP < 4\" \\\n",
    "--genotype-filter-name \"DP\" \\\n",
    "--set-filtered-genotype-to-no-call true \\\n",
    "--QUIET true\n",
    "Selecting variants\n",
    "\n",
    "gatk SelectVariants -R $REF \\\n",
    "-V CombineGenotypes/set_${set}_SNPs_VF.vcf.gz \\\n",
    "-O CombineGenotypes/set_${set}_SNPs_SV.vcf.gz \\\n",
    "--exclude-filtered true \\\n",
    "--exclude-non-variants true \\\n",
    "--remove-unused-alternates true \\\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 2. Filtering"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```\n",
    "TMP=../DATA_temp\n",
    "\n",
    "for bg in LL13 NC02\n",
    "  do\n",
    "  echo $bg\n",
    "\n",
    "  for type in SNPs INDELs\n",
    "      do\n",
    "\n",
    "      # Sample filters\n",
    "      # 1. Masking genotypes with DP < 4 (mask) and GT 2 or 3; \n",
    "      # 2. Allelic depth (AD) of the second most common variant within genotype is less than 4\n",
    "      # 3. Allelic depth of the second most frequent variant to most frequent variant is less than 0.2\n",
    "\n",
    "      python filterAD2_gatk.py ../DATA/set_${bg}_${type}_SV.vcf.gz\n",
    "      mv ../DATA/set_${bg}_${type}_SV_filterAD.vcf $TMP/set_${bg}_${type}_filterAD.vcf\n",
    "      bgzip -f $TMP/set_${bg}_${type}_filterAD.vcf .\n",
    "      tabix -f -p vcf $TMP/set_${bg}_${type}_filterAD.vcf.gz\n",
    "\n",
    "      # SNP filters\n",
    "      # 1. Mean per sample coverge is higher than 10\n",
    "      # 2. Max total DP is less than 20000\n",
    "      # 3. SNP QUAL is higher than 20\n",
    "      # 4. Mapping quality MQ is higher than 40\n",
    "      # 5. Fixed alleles removed (AF>0.99 or AC==0)\n",
    "\n",
    "      bcftools filter -e \"AVG(FORMAT/DP)<10 || INFO/DP>20000 || QUAL<20 || MQ<40 ||  (AF > 0.99) || (AC == 0)\" -Ou $TMP/set_${bg}_${type}_filterAD_tags.vcf.gz | bcftools view -m2 -Oz -o $TMP/set_${bg}_${type}_FLT1.vcf.gz -\n",
    "      bcftools stats $TMP/snp_bcftools_${bg}_FLT1.vcf.gz > snp_bcftools_${bg}_FLT1.stats\n",
    "      tabix -f -p vcf $TMP/snp_bcftools_${bg}_FLT1.vcf.gz\n",
    "      done\n",
    "  done\n",
    "```\n",
    "\n",
    "Python script available here: [filterAD2_gatk.py](filterAD2_gatk.py)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Annotating with snpEff and *S. cerevisiae* genome vR64-3-1 and splitting into SNPs and INDELs\n",
    "```\n",
    "conda activate snpeff-5.0\n",
    "# SnpEff version SnpEff 5.0e (build 2021-03-09 06:01), by Pablo Cingolani\n",
    "snpEff_DIR=/home/anna/soft/snpEff\n",
    "TMP=../DATA_temp\n",
    "CONFIG=/home/software/snpEff/snpEff.config\n",
    "\n",
    "for bg in LL13 NC02\n",
    "  do\n",
    "\n",
    "  for type in SNPs INDELs\n",
    "    do\n",
    "\n",
    "    # SnpEff\n",
    "    snpEff -c $CONFIG R64.3.1 $TMP/set_${bg}_${type}_FLT1.vcf.gz > $TMP/set_${bg}_${type}_snpEff.vcf\n",
    "    bgzip -f $TMP/set_${bg}_${type}_snpEff.vcf\n",
    "    tabix -f -p vcf $TMP/set_${bg}_${type}_snpEff.vcf.gz\n",
    "    cp $TMP/set_${bg}_${type}_snpEff.vcf* ../DATA/\n",
    "    done\n",
    "  done\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Getting a summary table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "important_categories = ['synonymous_variant','stop_gained','missense_variant']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Function for reading vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def readVcfGzip(filepath):\n",
    "    fh = gzip.open(filepath,'rt')\n",
    "    lines = [l for l in fh if not l.startswith('##')]\n",
    "    df_vcf = pd.read_csv(io.StringIO(''.join(lines)),sep='\\t')\n",
    "    return(df_vcf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Function for formatting vcf into stacked table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stackMutations(vcf_file):\n",
    "    \n",
    "    # Vcf to dataframe\n",
    "    df_vcf = readVcfGzip(vcf_file)\n",
    "    \n",
    "    # Formatting snpeff annotations & mutation\n",
    "    infos = df_vcf['INFO'].values.tolist()\n",
    "    I, G = [],[]\n",
    "    #T = []\n",
    "    for info in infos:\n",
    "        muts = []\n",
    "        ninfo = [i.split('=')[1] for i in info.split(';') if i.startswith('ANN=')][0]\n",
    "        sinfo = [i for i in ninfo.split(',')]\n",
    "        #T.append(ninfo)\n",
    "        mut = \":\".join([sinfo[0].split('|')[0],sinfo[0].split('|')[1]]), sinfo[0].split('|')[3].replace(\"_CDS\",\"\")\n",
    "        muts.append(mut)\n",
    "        for gene in sinfo[1:]:\n",
    "            ginfo = gene.split('|')\n",
    "            mut = \":\".join([ginfo[0],ginfo[1]]), ginfo[3].replace(\"_CDS\",\"\")\n",
    "            if (ginfo[1] in important_categories) & (mut not in muts):\n",
    "                muts.append(mut)\n",
    "        all_muts = '/'.join([m[0] for m in muts])\n",
    "        all_genes = '/'.join([m[1] for m in muts])\n",
    "        I.append(all_muts)\n",
    "        G.append(all_genes)\n",
    "    df_vcf['MUT'] = I\n",
    "    df_vcf['GENE'] = G\n",
    "    df_vcf['mutID'] = df_vcf.apply(lambda x: x[\"#CHROM\"]+\":\"+str(x[\"POS\"])+\":\"+str(x[\"REF\"])+\":\"+x[\"MUT\"], axis=1)\n",
    "    df_vcf.head()\n",
    "    \n",
    "    # Formatting samples\n",
    "    df_vcf_ = df_vcf.drop(columns=[\"ID\",\"FILTER\",\"INFO\",\"FORMAT\",\"MUT\"])\n",
    "    df_stack = df_vcf_.set_index([\"#CHROM\",\"POS\",\"REF\",\"ALT\",\"QUAL\",\"GENE\",\"mutID\"]).stack().reset_index().rename(columns={\"level_7\":\"sampleID\",0:\"GT\"})\n",
    "    del df_vcf_\n",
    "    df_stack['sample'] = df_stack['sampleID'].apply(lambda x: x.split('/')[-1].split('.')[0])\n",
    "    df_stack['snp'] = df_stack['GT'].apply(lambda x: 1 if x.split(':')[0] == '1' else 0)\n",
    "    df_stack = df_stack.drop(columns=[\"REF\",\"ALT\",\"QUAL\",\"sampleID\",\"GT\"])\n",
    "    df_stack = df_stack[df_stack['snp']==1].reset_index(drop=True)\n",
    "    del df_vcf\n",
    "    return(df_stack)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def stackMutationsWithAnn(vcf_file):\n",
    "    \n",
    "    # Vcf to dataframe\n",
    "    df_vcf = readVcfGzip(vcf_file)\n",
    "    \n",
    "    # Formatting snpeff annotations & mutation\n",
    "    infos = df_vcf['INFO'].values.tolist()\n",
    "    I, G = [],[]\n",
    "    #T = []\n",
    "    for info in infos:\n",
    "        muts = []\n",
    "        ninfo = [i.split('=')[1] for i in info.split(';') if i.startswith('ANN=')][0]\n",
    "        sinfo = [i for i in ninfo.split(',')]\n",
    "        #T.append(ninfo)\n",
    "        mut = \":\".join([sinfo[0].split('|')[0],sinfo[0].split('|')[1]]), sinfo[0].split('|')[3].replace(\"_CDS\",\"\")\n",
    "        muts.append(mut)\n",
    "        for gene in sinfo[1:]:\n",
    "            ginfo = gene.split('|')\n",
    "            mut = \":\".join([ginfo[0],ginfo[1]]), ginfo[3].replace(\"_CDS\",\"\")\n",
    "            if (ginfo[1] in important_categories) & (mut not in muts):\n",
    "                muts.append(mut)\n",
    "        all_muts = '/'.join([m[0] for m in muts])\n",
    "        all_genes = '/'.join([m[1] for m in muts])\n",
    "        I.append(all_muts)\n",
    "        G.append(all_genes)\n",
    "        \n",
    "    A = []\n",
    "    ANN = []\n",
    "    for info in infos:\n",
    "        muts = []\n",
    "        ninfo = [i.split('=')[1] for i in info.split(';') if i.startswith('ANN=')][0]\n",
    "        ANN.append(ninfo)\n",
    "        sinfo = [i for i in ninfo.split(',')]\n",
    "        #mut = \":\".join([sinfo[0].split('|')[0],sinfo[0].split('|')[1]]), sinfo[0].split('|')[3].replace(\"_CDS\",\"\")\n",
    "        #muts.append(mut)\n",
    "        for gene in sinfo:\n",
    "            ginfo = gene.split('|')\n",
    "            mut = \":\".join([ginfo[0],ginfo[1],ginfo[3].replace(\"_CDS\",\"\")])\n",
    "            if mut not in muts:\n",
    "                muts.append(mut)\n",
    "        all_muts = '/'.join(muts)\n",
    "        A.append(all_muts)\n",
    "    \n",
    "    df_vcf['MUT'] = I\n",
    "    df_vcf['GENE'] = G\n",
    "    df_vcf['MUT_ALL'] = A\n",
    "    df_vcf['ANN'] = ANN\n",
    "    df_vcf['mutID'] = df_vcf.apply(lambda x: x[\"#CHROM\"]+\":\"+str(x[\"POS\"])+\":\"+str(x[\"REF\"])+\":\"+x[\"MUT\"], axis=1)\n",
    "    df_vcf.head()\n",
    "    \n",
    "    # Formatting samples\n",
    "    df_vcf_ = df_vcf.drop(columns=[\"ID\",\"FILTER\",\"INFO\",\"FORMAT\",\"MUT\"])\n",
    "    df_stack = df_vcf_.set_index([\"#CHROM\",\"POS\",\"REF\",\"ALT\",\"QUAL\",\"GENE\",\"mutID\",\"MUT_ALL\",\"ANN\"]).stack().reset_index().rename(columns={\"level_9\":\"sampleID\",0:\"GT\"})\n",
    "    del df_vcf_\n",
    "    df_stack['sample'] = df_stack['sampleID'].apply(lambda x: x.split('/')[-1].split('.')[0])\n",
    "    df_stack['snp'] = df_stack['GT'].apply(lambda x: 1 if x.split(':')[0] == '1' else 0)\n",
    "    df_stack = df_stack.drop(columns=[\"QUAL\",\"sampleID\"])\n",
    "    df_stack = df_stack[df_stack['snp']==1].reset_index(drop=True)\n",
    "    del df_vcf\n",
    "    return(df_stack)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combining two backgrounds together"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27061, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,23:23:PASS:99:995,0</td>\n",
       "      <td>B10</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,22:22:PASS:99:974,0</td>\n",
       "      <td>B13</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,60:60:PASS:99:2628,0</td>\n",
       "      <td>B4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,55:55:PASS:99:2324,0</td>\n",
       "      <td>B5</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,46:46:PASS:99:1907,0</td>\n",
       "      <td>B6</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM  POS REF ALT     GENE                          mutID  \\\n",
       "0   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "1   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "2   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "3   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "4   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "1  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "2  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "3  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "4  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "\n",
       "                                                 ANN  \\\n",
       "0  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "1  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "2  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "3  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "4  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "\n",
       "                         GT sample  snp background     _merge  \n",
       "0   1:0,23:23:PASS:99:995,0    B10    1   LL13-040  left_only  \n",
       "1   1:0,22:22:PASS:99:974,0    B13    1   LL13-040  left_only  \n",
       "2  1:0,60:60:PASS:99:2628,0     B4    1   LL13-040  left_only  \n",
       "3  1:0,55:55:PASS:99:2324,0     B5    1   LL13-040  left_only  \n",
       "4  1:0,46:46:PASS:99:1907,0     B6    1   LL13-040  left_only  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf1 = \"../DATA/set_LL13_SNPs_snpEff.vcf.gz\"\n",
    "vcf2 = \"../DATA/set_NC02_SNPs_snpEff.vcf.gz\"\n",
    "# LL13\n",
    "df_vcf1 = stackMutationsWithAnn(vcf1)\n",
    "df_vcf1['background'] = \"LL13-040\"\n",
    "parent_variants_LL13 = df_vcf1.loc[df_vcf1['sample'].isin(['H13']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf1_merged = pd.merge(df_vcf1, parent_variants_LL13, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf1_noWT = df_vcf1_merged[df_vcf1_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "# NC02\n",
    "df_vcf2 = stackMutationsWithAnn(vcf2)\n",
    "df_vcf2['background'] = \"NC-02\"\n",
    "parent_variants_NC02 = df_vcf2.loc[df_vcf2['sample'].isin(['E23']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf2_merged = pd.merge(df_vcf2, parent_variants_NC02, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf2_noWT = df_vcf2_merged[df_vcf2_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "\n",
    "# Merged\n",
    "df = pd.concat([df_vcf1_noWT, df_vcf2_noWT],ignore_index=True)\n",
    "print(df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "df['NGENE'] = df['GENE'].apply(lambda x: list(set(x.split('/')))[0] if len(set(x.split('/'))) == 1 else x)\n",
    "#df[df['GENE'] == \"YHR128W/YHR128W\"].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Getting gene descriptions from SGD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Reading gene names extracted from SGD for corresponding gene IDs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Species</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>S000000001</td>\n",
       "      <td>YAL001C</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>TFC3</td>\n",
       "      <td>Transcription Factor class C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>S000000002</td>\n",
       "      <td>YAL002W</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>VPS8</td>\n",
       "      <td>Vacuolar Protein Sorting</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>S000000003</td>\n",
       "      <td>YAL003W</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>EFB1</td>\n",
       "      <td>Elongation Factor Beta</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>S000000004</td>\n",
       "      <td>YAL005C</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>SSA1</td>\n",
       "      <td>Stress-Seventy subfamily A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>S000000005</td>\n",
       "      <td>YAL007C</td>\n",
       "      <td>S. cerevisiae</td>\n",
       "      <td>ERP2</td>\n",
       "      <td>Emp24p/Erv25p Related Protein</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID     Gene        Species Symbol                           Name\n",
       "0  S000000001  YAL001C  S. cerevisiae   TFC3   Transcription Factor class C\n",
       "1  S000000002  YAL002W  S. cerevisiae   VPS8       Vacuolar Protein Sorting\n",
       "2  S000000003  YAL003W  S. cerevisiae   EFB1         Elongation Factor Beta\n",
       "3  S000000004  YAL005C  S. cerevisiae   SSA1     Stress-Seventy subfamily A\n",
       "4  S000000005  YAL007C  S. cerevisiae   ERP2  Emp24p/Erv25p Related Protein"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dg = pd.read_csv('../../DATA/S288C_reference_genome_R64-3-1_20210421/gene_list_SGD.tsv',sep=\"\\t\",names=['ID','Gene','Species','Symbol','Name'])\n",
    "dg.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Adding missing gene IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(27061, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,23:23:PASS:99:995,0</td>\n",
       "      <td>B10</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,22:22:PASS:99:974,0</td>\n",
       "      <td>B13</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,60:60:PASS:99:2628,0</td>\n",
       "      <td>B4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,55:55:PASS:99:2324,0</td>\n",
       "      <td>B5</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>353</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:353:C:T:missense_variant</td>\n",
       "      <td>T:missense_variant:YAL069W/T:upstream_gene_var...</td>\n",
       "      <td>T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...</td>\n",
       "      <td>1:0,46:46:PASS:99:1907,0</td>\n",
       "      <td>B6</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM  POS REF ALT     GENE                          mutID  \\\n",
       "0   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "1   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "2   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "3   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "4   chrI  353   C   T  YAL069W  chrI:353:C:T:missense_variant   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "1  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "2  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "3  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "4  T:missense_variant:YAL069W/T:upstream_gene_var...   \n",
       "\n",
       "                                                 ANN  \\\n",
       "0  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "1  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "2  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "3  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "4  T|missense_variant|MODERATE|YAL069W_CDS|GENE_Y...   \n",
       "\n",
       "                         GT sample  snp background     _merge    NGENE  \\\n",
       "0   1:0,23:23:PASS:99:995,0    B10    1   LL13-040  left_only  YAL069W   \n",
       "1   1:0,22:22:PASS:99:974,0    B13    1   LL13-040  left_only  YAL069W   \n",
       "2  1:0,60:60:PASS:99:2628,0     B4    1   LL13-040  left_only  YAL069W   \n",
       "3  1:0,55:55:PASS:99:2324,0     B5    1   LL13-040  left_only  YAL069W   \n",
       "4  1:0,46:46:PASS:99:1907,0     B6    1   LL13-040  left_only  YAL069W   \n",
       "\n",
       "  GENE_SYMBOL GENE_NAME  \n",
       "0         UNK       UNK  \n",
       "1         UNK       UNK  \n",
       "2         UNK       UNK  \n",
       "3         UNK       UNK  \n",
       "4         UNK       UNK  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dictG = {ele:np.array(ele.split('/')) for ele in df['NGENE'].drop_duplicates().values}\n",
    "all_gene_ids = np.concatenate(list(dictG.values()))\n",
    "missing_IDs = [ele for ele in all_gene_ids if ele not in dg['Gene'].values]\n",
    "missing_dg = pd.DataFrame({\"ID\":\"SXX\",\"Gene\":missing_IDs,\"Species\":\"S. cerevisiae\",\"Symbol\":\"UNK\",\"Name\":\"UNK\"})\n",
    "dg_combined = pd.concat([dg, missing_dg], ignore_index = True)\n",
    "\n",
    "gene_symbol = {a:b for a,b in dg_combined.get(['Gene','Symbol']).fillna('UNK').values.tolist()}\n",
    "gene_name = {a:b for a,b in dg_combined.get(['Gene','Name']).fillna('UNK').values.tolist()}\n",
    "\n",
    "df['GENE_SYMBOL'] = df['NGENE'].apply(lambda x: '/'.join([gene_symbol[i] for i in x.split('/')]))\n",
    "df['GENE_NAME'] = df['NGENE'].apply(lambda x: '/'.join([gene_name[i] for i in x.split('/')]))\n",
    "print(df.shape)\n",
    "df.to_csv('gatk_variants_SNP.tab',sep='\\t',header=True,index=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Grouping by snp - how many samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(5037, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1026</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1026:C:T:upstream_gene_variant</td>\n",
       "      <td>T:upstream_gene_variant:YAL067W-A/T:upstream_g...</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:0,15:15:PASS:99:555,0</td>\n",
       "      <td>E7</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1026</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1026:C:T:upstream_gene_variant</td>\n",
       "      <td>T:upstream_gene_variant:YAL067W-A/T:upstream_g...</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:1,5:6:PASS:99:122,0</td>\n",
       "      <td>F4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1026</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1026:C:T:upstream_gene_variant</td>\n",
       "      <td>T:upstream_gene_variant:YAL067W-A/T:upstream_g...</td>\n",
       "      <td>T|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:1,5:6:PASS:99:117,0</td>\n",
       "      <td>J3</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1232</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1232:T:C:upstream_gene_variant</td>\n",
       "      <td>C:upstream_gene_variant:YAL067W-A/C:upstream_g...</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:2,11:13:PASS:99:405,0</td>\n",
       "      <td>B16</td>\n",
       "      <td>1</td>\n",
       "      <td>NC-02</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1232</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1232:T:C:upstream_gene_variant</td>\n",
       "      <td>C:upstream_gene_variant:YAL067W-A/C:upstream_g...</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:1,10:11:PASS:99:425,0</td>\n",
       "      <td>B20</td>\n",
       "      <td>1</td>\n",
       "      <td>NC-02</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM   POS REF ALT       GENE                                mutID  \\\n",
       "0   chrI  1026   C   T  YAL067W-A  chrI:1026:C:T:upstream_gene_variant   \n",
       "1   chrI  1026   C   T  YAL067W-A  chrI:1026:C:T:upstream_gene_variant   \n",
       "2   chrI  1026   C   T  YAL067W-A  chrI:1026:C:T:upstream_gene_variant   \n",
       "3   chrI  1232   T   C  YAL067W-A  chrI:1232:T:C:upstream_gene_variant   \n",
       "4   chrI  1232   T   C  YAL067W-A  chrI:1232:T:C:upstream_gene_variant   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  T:upstream_gene_variant:YAL067W-A/T:upstream_g...   \n",
       "1  T:upstream_gene_variant:YAL067W-A/T:upstream_g...   \n",
       "2  T:upstream_gene_variant:YAL067W-A/T:upstream_g...   \n",
       "3  C:upstream_gene_variant:YAL067W-A/C:upstream_g...   \n",
       "4  C:upstream_gene_variant:YAL067W-A/C:upstream_g...   \n",
       "\n",
       "                                                 ANN                       GT  \\\n",
       "0  T|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...  1:0,15:15:PASS:99:555,0   \n",
       "1  T|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...    1:1,5:6:PASS:99:122,0   \n",
       "2  T|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...    1:1,5:6:PASS:99:117,0   \n",
       "3  C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...  1:2,11:13:PASS:99:405,0   \n",
       "4  C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...  1:1,10:11:PASS:99:425,0   \n",
       "\n",
       "  sample  snp background     _merge      NGENE GENE_SYMBOL GENE_NAME  \n",
       "0     E7    1   LL13-040  left_only  YAL067W-A         UNK       UNK  \n",
       "1     F4    1   LL13-040  left_only  YAL067W-A         UNK       UNK  \n",
       "2     J3    1   LL13-040  left_only  YAL067W-A         UNK       UNK  \n",
       "3    B16    1      NC-02  left_only  YAL067W-A         UNK       UNK  \n",
       "4    B20    1      NC-02  left_only  YAL067W-A         UNK       UNK  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dsamp = df.groupby([\"#CHROM\",\"POS\"]).agg(N_sample = (\"sample\",\"count\")).reset_index()\n",
    "dsamp_low = dsamp.loc[dsamp[\"N_sample\"] < 20, [\"#CHROM\",\"POS\"]]\n",
    "dff = pd.merge(dsamp_low, df, on = [\"#CHROM\",\"POS\"], how = \"left\")\n",
    "print(dff.shape)\n",
    "dff.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# INDELs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Combining two backgrounds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24216, 13)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>351</td>\n",
       "      <td>C</td>\n",
       "      <td>CATATAT</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:351:C:CATATAT:conservative_inframe_insertion</td>\n",
       "      <td>CATATAT:conservative_inframe_insertion:YAL069W...</td>\n",
       "      <td>CATATAT|conservative_inframe_insertion|MODERAT...</td>\n",
       "      <td>1:1,16:17:PASS:99:594,0</td>\n",
       "      <td>E6</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL069W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1400</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1400:CG:C:upstream_gene_variant</td>\n",
       "      <td>C:upstream_gene_variant:YAL067W-A/C:upstream_g...</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:0,5:5:PASS:99:225,0</td>\n",
       "      <td>O3</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1482</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1482:AC:A:upstream_gene_variant</td>\n",
       "      <td>A:upstream_gene_variant:YAL067W-A/A:upstream_g...</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:0,17:17:PASS:99:756,0</td>\n",
       "      <td>B4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2232</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:2232:AT:A:upstream_gene_variant</td>\n",
       "      <td>A:upstream_gene_variant:YAL068C/A:upstream_gen...</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...</td>\n",
       "      <td>1:1,18:19:PASS:99:765,0</td>\n",
       "      <td>B7</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2232</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:2232:AT:A:upstream_gene_variant</td>\n",
       "      <td>A:upstream_gene_variant:YAL068C/A:upstream_gen...</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...</td>\n",
       "      <td>1:3,18:21:PASS:99:675,0</td>\n",
       "      <td>C3</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM   POS REF      ALT       GENE  \\\n",
       "0   chrI   351   C  CATATAT    YAL069W   \n",
       "1   chrI  1400  CG        C  YAL067W-A   \n",
       "2   chrI  1482  AC        A  YAL067W-A   \n",
       "3   chrI  2232  AT        A    YAL068C   \n",
       "4   chrI  2232  AT        A    YAL068C   \n",
       "\n",
       "                                               mutID  \\\n",
       "0  chrI:351:C:CATATAT:conservative_inframe_insertion   \n",
       "1               chrI:1400:CG:C:upstream_gene_variant   \n",
       "2               chrI:1482:AC:A:upstream_gene_variant   \n",
       "3               chrI:2232:AT:A:upstream_gene_variant   \n",
       "4               chrI:2232:AT:A:upstream_gene_variant   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  CATATAT:conservative_inframe_insertion:YAL069W...   \n",
       "1  C:upstream_gene_variant:YAL067W-A/C:upstream_g...   \n",
       "2  A:upstream_gene_variant:YAL067W-A/A:upstream_g...   \n",
       "3  A:upstream_gene_variant:YAL068C/A:upstream_gen...   \n",
       "4  A:upstream_gene_variant:YAL068C/A:upstream_gen...   \n",
       "\n",
       "                                                 ANN                       GT  \\\n",
       "0  CATATAT|conservative_inframe_insertion|MODERAT...  1:1,16:17:PASS:99:594,0   \n",
       "1  C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...    1:0,5:5:PASS:99:225,0   \n",
       "2  A|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...  1:0,17:17:PASS:99:756,0   \n",
       "3  A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...  1:1,18:19:PASS:99:765,0   \n",
       "4  A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...  1:3,18:21:PASS:99:675,0   \n",
       "\n",
       "  sample  snp background     _merge      NGENE  \n",
       "0     E6    1   LL13-040  left_only    YAL069W  \n",
       "1     O3    1   LL13-040  left_only  YAL067W-A  \n",
       "2     B4    1   LL13-040  left_only  YAL067W-A  \n",
       "3     B7    1   LL13-040  left_only    YAL068C  \n",
       "4     C3    1   LL13-040  left_only    YAL068C  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf1 = \"../DATA/set_LL13_INDELs_snpEff.vcf.gz\"\n",
    "vcf2 = \"../DATA/set_NC02_INDELs_snpEff.vcf.gz\"\n",
    "# LL13\n",
    "df_vcf1 = stackMutationsWithAnn(vcf1)\n",
    "df_vcf1['background'] = \"LL13-040\"\n",
    "parent_variants_LL13 = df_vcf1.loc[df_vcf1['sample'].isin(['H13']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf1_merged = pd.merge(df_vcf1, parent_variants_LL13, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf1_noWT = df_vcf1_merged[df_vcf1_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "# NC02\n",
    "df_vcf2 = stackMutationsWithAnn(vcf2)\n",
    "df_vcf2['background'] = \"NC-02\"\n",
    "parent_variants_NC02 = df_vcf2.loc[df_vcf2['sample'].isin(['E23']), ['#CHROM','POS']].reset_index(drop=True)\n",
    "df_vcf2_merged = pd.merge(df_vcf2, parent_variants_NC02, on = ['#CHROM','POS'], how = 'outer', indicator=True)\n",
    "df_vcf2_noWT = df_vcf2_merged[df_vcf2_merged['_merge'] == 'left_only'].reset_index(drop=True)\n",
    "\n",
    "# Merged\n",
    "df = pd.concat([df_vcf1_noWT, df_vcf2_noWT],ignore_index=True)\n",
    "print(df.shape)\n",
    "\n",
    "# Gene name\n",
    "df['NGENE'] = df['GENE'].apply(lambda x: list(set(x.split('/')))[0] if len(set(x.split('/'))) == 1 else x)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6040</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362176</td>\n",
       "      <td>TG</td>\n",
       "      <td>T</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362176:TG:T:frameshift_variant</td>\n",
       "      <td>T:frameshift_variant:YHR128W/T:upstream_gene_v...</td>\n",
       "      <td>T|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...</td>\n",
       "      <td>1:0,24:24:PASS:99:947,0</td>\n",
       "      <td>B11</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6041</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362273</td>\n",
       "      <td>T</td>\n",
       "      <td>TCTACCTGTGCAAAAGCAAATTGTGGAAACTGACACCAA</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362273:T:TCTACCTGTGCAAAAGCAAATTGTGGAAA...</td>\n",
       "      <td>TCTACCTGTGCAAAAGCAAATTGTGGAAACTGACACCAA:frames...</td>\n",
       "      <td>TCTACCTGTGCAAAAGCAAATTGTGGAAACTGACACCAA|frames...</td>\n",
       "      <td>1:0,41:41:PASS:99:1828,0</td>\n",
       "      <td>K7</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6042</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362550</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362550:GC:G:frameshift_variant</td>\n",
       "      <td>G:frameshift_variant:YHR128W/G:upstream_gene_v...</td>\n",
       "      <td>G|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...</td>\n",
       "      <td>1:0,14:14:PASS:99:567,0</td>\n",
       "      <td>F11</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6043</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362661</td>\n",
       "      <td>CCAGAGGTCAGAATTGTTACTGGTGCCCTCGA</td>\n",
       "      <td>C</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362661:CCAGAGGTCAGAATTGTTACTGGTGCCCTCG...</td>\n",
       "      <td>C:frameshift_variant:YHR128W/C:upstream_gene_v...</td>\n",
       "      <td>C|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...</td>\n",
       "      <td>1:0,34:34:PASS:99:1535,0</td>\n",
       "      <td>E13</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6044</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362725</td>\n",
       "      <td>CA</td>\n",
       "      <td>C</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362725:CA:C:frameshift_variant</td>\n",
       "      <td>C:frameshift_variant:YHR128W/C:upstream_gene_v...</td>\n",
       "      <td>C|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...</td>\n",
       "      <td>1:0,7:7:PASS:99:283,0</td>\n",
       "      <td>B13</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17755</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362246</td>\n",
       "      <td>ATTGT</td>\n",
       "      <td>A</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362246:ATTGT:A:frameshift_variant</td>\n",
       "      <td>A:frameshift_variant:YHR128W/A:upstream_gene_v...</td>\n",
       "      <td>A|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...</td>\n",
       "      <td>1:1,23:24:99:990,0</td>\n",
       "      <td>J15</td>\n",
       "      <td>1</td>\n",
       "      <td>NC-02</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17756</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362254</td>\n",
       "      <td>TTGA</td>\n",
       "      <td>T</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362254:TTGA:T:disruptive_inframe_deletion</td>\n",
       "      <td>T:disruptive_inframe_deletion:YHR128W/T:upstre...</td>\n",
       "      <td>T|disruptive_inframe_deletion|MODERATE|YHR128W...</td>\n",
       "      <td>1:1,23:24:99:990,0</td>\n",
       "      <td>J15</td>\n",
       "      <td>1</td>\n",
       "      <td>NC-02</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17757</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362282</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362282:GC:G:frameshift_variant</td>\n",
       "      <td>G:frameshift_variant:YHR128W/G:upstream_gene_v...</td>\n",
       "      <td>G|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...</td>\n",
       "      <td>1:0,29:29:PASS:99:1143,0</td>\n",
       "      <td>K19</td>\n",
       "      <td>1</td>\n",
       "      <td>NC-02</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17758</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362360</td>\n",
       "      <td>CATTGTCAGAGCTGGT</td>\n",
       "      <td>C</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362360:CATTGTCAGAGCTGGT:C:conservative...</td>\n",
       "      <td>C:conservative_inframe_deletion:YHR128W/C:upst...</td>\n",
       "      <td>C|conservative_inframe_deletion|MODERATE|YHR12...</td>\n",
       "      <td>1:0,29:29:PASS:99:1314,0</td>\n",
       "      <td>B20</td>\n",
       "      <td>1</td>\n",
       "      <td>NC-02</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17759</th>\n",
       "      <td>chrVIII</td>\n",
       "      <td>362378</td>\n",
       "      <td>ATCGATGGAGCAAGGATTAAGAGACTGTTGTAGGTCTGTGCGTATCGGT</td>\n",
       "      <td>A</td>\n",
       "      <td>YHR128W</td>\n",
       "      <td>chrVIII:362378:ATCGATGGAGCAAGGATTAAGAGACTGTTGT...</td>\n",
       "      <td>A:conservative_inframe_deletion:YHR128W/A:upst...</td>\n",
       "      <td>A|conservative_inframe_deletion|MODERATE|YHR12...</td>\n",
       "      <td>1:0,29:29:PASS:99:1317,0</td>\n",
       "      <td>B20</td>\n",
       "      <td>1</td>\n",
       "      <td>NC-02</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YHR128W</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        #CHROM     POS                                                REF  \\\n",
       "6040   chrVIII  362176                                                 TG   \n",
       "6041   chrVIII  362273                                                  T   \n",
       "6042   chrVIII  362550                                                 GC   \n",
       "6043   chrVIII  362661                   CCAGAGGTCAGAATTGTTACTGGTGCCCTCGA   \n",
       "6044   chrVIII  362725                                                 CA   \n",
       "17755  chrVIII  362246                                              ATTGT   \n",
       "17756  chrVIII  362254                                               TTGA   \n",
       "17757  chrVIII  362282                                                 GC   \n",
       "17758  chrVIII  362360                                   CATTGTCAGAGCTGGT   \n",
       "17759  chrVIII  362378  ATCGATGGAGCAAGGATTAAGAGACTGTTGTAGGTCTGTGCGTATCGGT   \n",
       "\n",
       "                                           ALT     GENE  \\\n",
       "6040                                         T  YHR128W   \n",
       "6041   TCTACCTGTGCAAAAGCAAATTGTGGAAACTGACACCAA  YHR128W   \n",
       "6042                                         G  YHR128W   \n",
       "6043                                         C  YHR128W   \n",
       "6044                                         C  YHR128W   \n",
       "17755                                        A  YHR128W   \n",
       "17756                                        T  YHR128W   \n",
       "17757                                        G  YHR128W   \n",
       "17758                                        C  YHR128W   \n",
       "17759                                        A  YHR128W   \n",
       "\n",
       "                                                   mutID  \\\n",
       "6040              chrVIII:362176:TG:T:frameshift_variant   \n",
       "6041   chrVIII:362273:T:TCTACCTGTGCAAAAGCAAATTGTGGAAA...   \n",
       "6042              chrVIII:362550:GC:G:frameshift_variant   \n",
       "6043   chrVIII:362661:CCAGAGGTCAGAATTGTTACTGGTGCCCTCG...   \n",
       "6044              chrVIII:362725:CA:C:frameshift_variant   \n",
       "17755          chrVIII:362246:ATTGT:A:frameshift_variant   \n",
       "17756  chrVIII:362254:TTGA:T:disruptive_inframe_deletion   \n",
       "17757             chrVIII:362282:GC:G:frameshift_variant   \n",
       "17758  chrVIII:362360:CATTGTCAGAGCTGGT:C:conservative...   \n",
       "17759  chrVIII:362378:ATCGATGGAGCAAGGATTAAGAGACTGTTGT...   \n",
       "\n",
       "                                                 MUT_ALL  \\\n",
       "6040   T:frameshift_variant:YHR128W/T:upstream_gene_v...   \n",
       "6041   TCTACCTGTGCAAAAGCAAATTGTGGAAACTGACACCAA:frames...   \n",
       "6042   G:frameshift_variant:YHR128W/G:upstream_gene_v...   \n",
       "6043   C:frameshift_variant:YHR128W/C:upstream_gene_v...   \n",
       "6044   C:frameshift_variant:YHR128W/C:upstream_gene_v...   \n",
       "17755  A:frameshift_variant:YHR128W/A:upstream_gene_v...   \n",
       "17756  T:disruptive_inframe_deletion:YHR128W/T:upstre...   \n",
       "17757  G:frameshift_variant:YHR128W/G:upstream_gene_v...   \n",
       "17758  C:conservative_inframe_deletion:YHR128W/C:upst...   \n",
       "17759  A:conservative_inframe_deletion:YHR128W/A:upst...   \n",
       "\n",
       "                                                     ANN  \\\n",
       "6040   T|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...   \n",
       "6041   TCTACCTGTGCAAAAGCAAATTGTGGAAACTGACACCAA|frames...   \n",
       "6042   G|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...   \n",
       "6043   C|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...   \n",
       "6044   C|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...   \n",
       "17755  A|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...   \n",
       "17756  T|disruptive_inframe_deletion|MODERATE|YHR128W...   \n",
       "17757  G|frameshift_variant|HIGH|YHR128W_CDS|GENE_YHR...   \n",
       "17758  C|conservative_inframe_deletion|MODERATE|YHR12...   \n",
       "17759  A|conservative_inframe_deletion|MODERATE|YHR12...   \n",
       "\n",
       "                             GT sample  snp background     _merge    NGENE  \n",
       "6040    1:0,24:24:PASS:99:947,0    B11    1   LL13-040  left_only  YHR128W  \n",
       "6041   1:0,41:41:PASS:99:1828,0     K7    1   LL13-040  left_only  YHR128W  \n",
       "6042    1:0,14:14:PASS:99:567,0    F11    1   LL13-040  left_only  YHR128W  \n",
       "6043   1:0,34:34:PASS:99:1535,0    E13    1   LL13-040  left_only  YHR128W  \n",
       "6044      1:0,7:7:PASS:99:283,0    B13    1   LL13-040  left_only  YHR128W  \n",
       "17755        1:1,23:24:99:990,0    J15    1      NC-02  left_only  YHR128W  \n",
       "17756        1:1,23:24:99:990,0    J15    1      NC-02  left_only  YHR128W  \n",
       "17757  1:0,29:29:PASS:99:1143,0    K19    1      NC-02  left_only  YHR128W  \n",
       "17758  1:0,29:29:PASS:99:1314,0    B20    1      NC-02  left_only  YHR128W  \n",
       "17759  1:0,29:29:PASS:99:1317,0    B20    1      NC-02  left_only  YHR128W  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[[\"#CHROM\",\"POS\"]].drop_duplicates()\n",
    "df[df['NGENE'] == \"YHR128W\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### SGD gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(24216, 16)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>GENE</th>\n",
       "      <th>mutID</th>\n",
       "      <th>MUT_ALL</th>\n",
       "      <th>ANN</th>\n",
       "      <th>GT</th>\n",
       "      <th>sample</th>\n",
       "      <th>snp</th>\n",
       "      <th>background</th>\n",
       "      <th>_merge</th>\n",
       "      <th>NGENE</th>\n",
       "      <th>GENE_SYMBOL</th>\n",
       "      <th>GENE_NAME</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>351</td>\n",
       "      <td>C</td>\n",
       "      <td>CATATAT</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>chrI:351:C:CATATAT:conservative_inframe_insertion</td>\n",
       "      <td>CATATAT:conservative_inframe_insertion:YAL069W...</td>\n",
       "      <td>CATATAT|conservative_inframe_insertion|MODERAT...</td>\n",
       "      <td>1:1,16:17:PASS:99:594,0</td>\n",
       "      <td>E6</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1400</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1400:CG:C:upstream_gene_variant</td>\n",
       "      <td>C:upstream_gene_variant:YAL067W-A/C:upstream_g...</td>\n",
       "      <td>C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:0,5:5:PASS:99:225,0</td>\n",
       "      <td>O3</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>1482</td>\n",
       "      <td>AC</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>chrI:1482:AC:A:upstream_gene_variant</td>\n",
       "      <td>A:upstream_gene_variant:YAL067W-A/A:upstream_g...</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...</td>\n",
       "      <td>1:0,17:17:PASS:99:756,0</td>\n",
       "      <td>B4</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>UNK</td>\n",
       "      <td>UNK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2232</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:2232:AT:A:upstream_gene_variant</td>\n",
       "      <td>A:upstream_gene_variant:YAL068C/A:upstream_gen...</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...</td>\n",
       "      <td>1:1,18:19:PASS:99:765,0</td>\n",
       "      <td>B7</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>seriPAUperin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>2232</td>\n",
       "      <td>AT</td>\n",
       "      <td>A</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>chrI:2232:AT:A:upstream_gene_variant</td>\n",
       "      <td>A:upstream_gene_variant:YAL068C/A:upstream_gen...</td>\n",
       "      <td>A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...</td>\n",
       "      <td>1:3,18:21:PASS:99:675,0</td>\n",
       "      <td>C3</td>\n",
       "      <td>1</td>\n",
       "      <td>LL13-040</td>\n",
       "      <td>left_only</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>PAU8</td>\n",
       "      <td>seriPAUperin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  #CHROM   POS REF      ALT       GENE  \\\n",
       "0   chrI   351   C  CATATAT    YAL069W   \n",
       "1   chrI  1400  CG        C  YAL067W-A   \n",
       "2   chrI  1482  AC        A  YAL067W-A   \n",
       "3   chrI  2232  AT        A    YAL068C   \n",
       "4   chrI  2232  AT        A    YAL068C   \n",
       "\n",
       "                                               mutID  \\\n",
       "0  chrI:351:C:CATATAT:conservative_inframe_insertion   \n",
       "1               chrI:1400:CG:C:upstream_gene_variant   \n",
       "2               chrI:1482:AC:A:upstream_gene_variant   \n",
       "3               chrI:2232:AT:A:upstream_gene_variant   \n",
       "4               chrI:2232:AT:A:upstream_gene_variant   \n",
       "\n",
       "                                             MUT_ALL  \\\n",
       "0  CATATAT:conservative_inframe_insertion:YAL069W...   \n",
       "1  C:upstream_gene_variant:YAL067W-A/C:upstream_g...   \n",
       "2  A:upstream_gene_variant:YAL067W-A/A:upstream_g...   \n",
       "3  A:upstream_gene_variant:YAL068C/A:upstream_gen...   \n",
       "4  A:upstream_gene_variant:YAL068C/A:upstream_gen...   \n",
       "\n",
       "                                                 ANN                       GT  \\\n",
       "0  CATATAT|conservative_inframe_insertion|MODERAT...  1:1,16:17:PASS:99:594,0   \n",
       "1  C|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...    1:0,5:5:PASS:99:225,0   \n",
       "2  A|upstream_gene_variant|MODIFIER|YAL067W-A_CDS...  1:0,17:17:PASS:99:756,0   \n",
       "3  A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...  1:1,18:19:PASS:99:765,0   \n",
       "4  A|upstream_gene_variant|MODIFIER|YAL068C_CDS|G...  1:3,18:21:PASS:99:675,0   \n",
       "\n",
       "  sample  snp background     _merge      NGENE GENE_SYMBOL     GENE_NAME  \n",
       "0     E6    1   LL13-040  left_only    YAL069W         UNK           UNK  \n",
       "1     O3    1   LL13-040  left_only  YAL067W-A         UNK           UNK  \n",
       "2     B4    1   LL13-040  left_only  YAL067W-A         UNK           UNK  \n",
       "3     B7    1   LL13-040  left_only    YAL068C        PAU8  seriPAUperin  \n",
       "4     C3    1   LL13-040  left_only    YAL068C        PAU8  seriPAUperin  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dg = pd.read_csv('../../DATA/S288C_reference_genome_R64-3-1_20210421/gene_list_SGD.tsv',sep=\"\\t\",names=['ID','Gene','Species','Symbol','Name'])\n",
    "dictG = {ele:np.array(ele.split('/')) for ele in df['NGENE'].drop_duplicates().values}\n",
    "all_gene_ids = np.concatenate(list(dictG.values()))\n",
    "missing_IDs = [ele for ele in all_gene_ids if ele not in dg['Gene'].values]\n",
    "missing_dg = pd.DataFrame({\"ID\":\"SXX\",\"Gene\":missing_IDs,\"Species\":\"S. cerevisiae\",\"Symbol\":\"UNK\",\"Name\":\"UNK\"})\n",
    "dg_combined = pd.concat([dg, missing_dg], ignore_index = True)\n",
    "\n",
    "gene_symbol = {a:b for a,b in dg_combined.get(['Gene','Symbol']).fillna('UNK').values.tolist()}\n",
    "gene_name = {a:b for a,b in dg_combined.get(['Gene','Name']).fillna('UNK').values.tolist()}\n",
    "\n",
    "df['GENE_SYMBOL'] = df['NGENE'].apply(lambda x: '/'.join([gene_symbol[i] for i in x.split('/')]))\n",
    "df['GENE_NAME'] = df['NGENE'].apply(lambda x: '/'.join([gene_name[i] for i in x.split('/')]))\n",
    "print(df.shape)\n",
    "df.to_csv('gatk_variants_INDEL.tab',sep='\\t',header=True,index=True)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
